gcs/gcs.go

```diff
@@ -0,0 +1,348 @@
+func (f *Filter) N() uint32 {
+	return f.n
+}
+
+// Match checks whether a []byte value is likely (within collision probability)
+// to be a member of the set represented by the filter.
+func (f *Filter) Match(key [KeySize]byte, data []byte) bool {
+	// Create a filter bitstream.
+	b := newBitReader(f.Bytes())
+
+	// Hash our search term with the same parameters as the filter.
+	term := siphash.Sum64(data, &key) % f.modulusNP
+
+	// Go through the search filter and look for the desired value.
+	var lastValue uint64
+	for lastValue < term {
+		// Read the difference between previous and new value from
+		// bitstream.
+		value, err := f.readFullUint64(&b)
+		if err != nil {
+			return false
+		}
+
+		// Add the previous value to it.
+		value += lastValue
+		if value == term {
+			return true
+		}
+
+		lastValue = value
+	}
+
+	return false
+}
+
+// MatchAny returns checks whether any []byte value is likely (within collision
+// probability) to be a member of the set represented by the filter faster than
+// calling Match() for each value individually.
+func (f *Filter) MatchAny(key [KeySize]byte, data [][]byte) bool {
+	if len(data) == 0 {
+		return false
+	}
+
+	// Create a filter bitstream.
+	b := newBitReader(f.Bytes())
+
+	// Create an uncompressed filter of the search values.
+	values := make(uint64Slice, 0, len(data))
+	for _, d := range data {
+		v := siphash.Sum64(d, &key) % f.modulusNP
+		values = append(values, v)
+	}
+	sort.Sort(values)
+
+	// Zip down the filters, comparing values until we either run out of
+	// values to compare in one of the filters or we reach a matching
+	// value.
+	var lastValue1, lastValue2 uint64
+	lastValue2 = values[0]
+	i := 1
+	for lastValue1 != lastValue2 {
+		// Check which filter to advance to make sure we're comparing
+		// the right values.
+		switch {
+		case lastValue1 > lastValue2:
+			// Advance filter created from search terms or return
+			// false if we're at the end because nothing matched.
+			if i < len(values) {
+				lastValue2 = values[i]
+				i++
+			} else {
+				return false
+			}
+		case lastValue2 > lastValue1:
+			// Advance filter we're searching or return false if
+			// we're at the end because nothing matched.
+			value, err := f.readFullUint64(&b)
+			if err != nil {
+				return false
+			}
+			lastValue1 += value
+		}
+	}
+
+	// If we've made it this far, an element matched between filters so we
+	// return true.
+	return true
+}
+
+// readFullUint64 reads a value represented by the sum of a unary multiple of
+// the filter's P modulus (`2**P`) and a big-endian P-bit remainder.
+func (f *Filter) readFullUint64(b *bitReader) (uint64, error) {
+	v, err := b.readUnary()
+	if err != nil {
+		return 0, err
+	}
+
+	rem, err := b.readNBits(uint(f.p))
+	if err != nil {
+		return 0, err
+	}
+
+	// Add the multiple and the remainder.
+	return v*f.modulusP + rem, nil
+}
+
+// Hash returns the double-BLAKE256 hash of the filter.
```

Comment:
We should change this to be a single hash.  There is no reason to double hash in BLAKE256 since the hash function isn't vulnerable to length extension attacks which is why double sha256 is used in the upstream code.